### DISCUSSION TITLE
Fetcher functions seems to be cached by useSWR(?)

### DISCUSSION THREAD

**Discussion by westmark** [vercel/swr#1805] [Q&A] [answered]
_v 1.1.2_

Hi! I have the weirdest bug. Consider this custom hook

```typescript
const useCompanyTeams = (
  companyId: Maybe<string>,
  config: SWRConfiguration = {},
) => {
  const api = useApi<CompaniesApi>('CompaniesApi');

  const rand = Math.random().toString(32).substring(2);
  console.log(rand, companyId, companyId ? companyTeamsKey(companyId) : null);

  return useSWR<Array<Team>>(
    companyId ? companyTeamsKey(companyId) : null,
    (key) => {
      console.log(
        'Calling',
        rand,
        companyId ? companyTeamsKey(companyId) : null,
        key,
      );
      return getCompanyTeams(companyId, api);
    },
    {
      suspense: true,
      ...config,
    },
  );
};
```

(`useApi` returns a generated API with all paths already baked in, and the random value is only for debugging purposes)

During page load, this hook might be called with `companyId` set to `null`, which should yield a null key leading to no data being fetched. Here is the output from the application

```
> 39ftoqmq8ng null null
> 9f0aol3gcgg undefined null
> hass14fndqg null null
> msf3pvqppk8 e6f1a6ba-febf-433c-bb94-eeeae26f6e6b /companies/e6f1a6ba-febf-433c-bb94-eeeae26f6e6b/teams
> calling bl8s3lb5j2o null  /companies/e6f1a6ba-febf-433c-bb94-eeeae26f6e6b/teams
```

The last line above is interesting. 

1. The `rand` variable is logged immediately after its creation, but `bl8s3lb5j2o` which is printed INSIDE the fetcher closure does not appear above, which it should. 
2. Furthermore, `companyId` is `null` inside the closure, but the `key` argument is valid. 

I can't explain 1) at all. This only happens if the hook is invoked with `companyId` set to null before it becomes non-null. If invoked with a non-null value the first time, everything will work and the log will be what your would expect. 

2) seems to suggest that the fetcher(i.e. the closure) is cached inside SWR. Is it?

Can anyone shine some light on this?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Normally, the fetcher is refreshed through an isomorphic `useEffect`, but I reckon you have `suspense: true`. Could you help me build a reproducible demo?

```js
import { Suspense, useState } from "react";
import useSWR from "swr";
import "./styles.css";

const useData = (str) => {
  const ext = Math.random();
  console.log("outer", { ext, str });

  return useSWR(
    str,
    () => {
      console.log("inner", { ext, str });
      return new Promise((resolve) => setTimeout(resolve, 500, ext));
    },
    { suspense: true }
  );
};

function Main({ text }) {
  useData(text || null);

  return <p>{text}</p>;
}

export default function App() {
  const [text, setText] = useState("");
  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>

      <input
        type="text"
        value={text}
        onChange={(e) => setText(e.target.value)}
      />
      <Suspense fallback={null}>
        <Main text={text} />
      </Suspense>
    </div>
  );
}
```

Here's a codesandbox that runs the above, https://codesandbox.io/s/hungry-monad-9w8ok?file=/src/App.js, when you type on letter into the input, are the logs reflecting the same problem?

### What I see

When `suspense` is set `true`, there seems to be a stale version of the fetcher, which is immediately fixed on the next render pass.

Could you try to edit/review that codesandbox?

**Reply by westmark** [reply]
@icyJoseph Yes it seems to be the same issue. The problem for me is that the underlying api call throws when the in-data is bad, so we never got to see the fetcher fix itself. Anyway, I fixed this by adding the relevant data to the key as an array, so I can get the actual value instead of the closed over version. Still perplexing though

**Reply by icyJoseph** [reply]
I think I have a fix for this, I've made an issue out of this discussion and submitted a PR.

**Reply by icyJoseph** [reply]
Actually, #1803 was submitted 2 days ago, and it is proposing a more scoped (for suspense only) solution.

**Reply by icyJoseph** [reply]
Follow up #1803 has been merged into main, and should be out in the next release. 

Meaning that, after the next release, the fetcher function should always be correct, under suspense mode.
